app/cluster/dynamic.go

```diff
@@ -62,33 +73,60 @@ func (d *Dynamic) init() {
 	d.serverStopTimeStat = stats.NewTaggedStat("cluster.server_stop_time", stats.TimerType, tag)
 	d.serverStartCountStat = stats.NewTaggedStat("cluster.server_start_count", stats.CountType, tag)
 	d.serverStopCountStat = stats.NewTaggedStat("cluster.server_stop_count", stats.CountType, tag)
+
+	if d.BackendConfig == nil {
+		d.BackendConfig = backendconfig.DefaultBackendConfig
+	}
 }
 
 func (d *Dynamic) Run(ctx context.Context) error {
-	d.once.Do(func() {
-		d.init()
-	})
-	defer d.Provider.Close()
-	modeChan, err := d.Provider.ServerMode()
-	if err != nil {
-		return err
-	}
+	d.once.Do(d.init)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	serverModeChan := d.Provider.ServerMode(ctx)
+	workspaceIDsChan := d.Provider.WorkspaceIDs(ctx)
+
 	for {
 		select {
 		case <-ctx.Done():
 			if d.currentMode == servermode.NormalMode {
 				d.stop()
 			}
 			return nil
```

Comment:
Wouldn't it make sense to return `ctx.Err()` instead?

Reply:
Not sure about that, `ctx` cancellation is the expected way to terminate this component I don't see any benefit in returning an error.

Reply:
> I don't see any benefit in returning an error.

That's just because nothing inside actually honours the context (e.g. `req.Ack()` is not `req.Ack(ctx)`).
Let's say that then we add one. They would propagate the error like most libraries do when a context is canceled so that then on the other side you can do stuff like `errors.Is(err, context.DeadlineExceeded)` and act accordingly (if you want to).

If you don't return it I think you might just be complicating things so that in some cases it is propagated and in some others not. Returning it is the most obvious (and expected) thing to do in my opinion although the problem might not present itself now, this code might be changed by someone else in the future and not consider this.

It is up to you but keep in mind that sometimes not doing the obvious thing is a cost that you might have to pay later.